trie/snap_diskcache.go

```diff
@@ -0,0 +1,204 @@
+	if ok {
+		if n.hash != hash {
+			owner, path := DecodeStorageKey(storage)
+			return nil, fmt.Errorf("%w %x(%x %v)", errUnexpectedNode, hash, owner, path)
+		}
+		triedbDirtyHitMeter.Mark(1)
+		triedbDirtyNodeHitDepthHist.Update(int64(128))
+		triedbDirtyReadMeter.Mark(int64(n.size))
+		return n, nil
+	}
+	return nil, nil
+}
+
+// commit merges the given dirty nodes into the cache and bumps seq as well
+// to complete the state transition.
+func (cache *diskcache) commit(nodes map[string]*cachedNode) *diskcache {
+	cache.seq += 1
+
+	var size int64
+	for storage, node := range nodes {
+		if orig, exist := cache.nodes[storage]; !exist {
+			size += int64(node.memorySize(len(storage)))
+		} else {
+			size += int64(node.size) - int64(orig.size)
+			triedbGCNodesMeter.Mark(1)
+			triedbGCSizeMeter.Mark(int64(orig.size))
+		}
+		cache.nodes[storage] = node
+	}
+	if final := int64(cache.size) + size; final < 0 {
+		log.Error("Negative disk cache size", "previous", common.StorageSize(cache.size), "diff", common.StorageSize(size))
+		cache.size = 0
+	} else {
+		cache.size = uint64(final)
+	}
+	return cache
+}
+
+// revert applies the reverse diff to the local dirty node set.
+func (cache *diskcache) revert(diff *reverseDiff) error {
+	if cache.seq == 0 {
+		return errStateUnrecoverable
+	}
+	cache.seq -= 1
+
+	// If all the embedded state transitions are reverted,
+	// reset the cache entirely.
+	if cache.seq == 0 {
+		cache.reset()
+		return nil
+	}
+	for _, state := range diff.States {
+		_, ok := cache.nodes[string(state.Key)]
+		if !ok {
+			// TODO it should never happen, perhaps panic here.
+			owner, path := DecodeStorageKey(state.Key)
+			return fmt.Errorf("non-existent node (%x %v)", owner, path)
+		}
+		if len(state.Val) == 0 {
+			cache.nodes[string(state.Key)] = &cachedNode{
+				node: nil,
+				size: 0,
+				hash: common.Hash{},
+			}
+		} else {
+			cache.nodes[string(state.Key)] = &cachedNode{
+				node: rawNode(state.Val),
+				size: uint16(len(state.Val)),
+				hash: crypto.Keccak256Hash(state.Val),
+			}
+		}
+	}
+	return nil
+}
+
+// reset cleans up the disk cache.
+func (cache *diskcache) reset() {
+	cache.seq = 0
+	cache.nodes = make(map[string]*cachedNode)
+	cache.size = 0
+}
+
+// empty returns an indicator if diskcache contains any state transition inside.
+func (cache *diskcache) empty() bool {
+	return cache.seq == 0
+}
+
+// forEach iterates all the cached nodes and applies the given callback on them
+func (cache *diskcache) forEach(callback func(key string, node *cachedNode)) {
+	for storage, n := range cache.nodes {
+		callback(storage, n)
+	}
+}
+
+// mayFlush persists the in-memory dirty trie node into the disk if the predefined
+// memory threshold is reached. Note, all data must be written to disk atomically.
+// This function should never be called simultaneously with other map accessors.
+func (cache *diskcache) mayFlush(db ethdb.KeyValueStore, clean *fastcache.Cache, diffid uint64, force bool) error {
```

Comment:
There's an interesting race that can happen here: the clean cache - if it's accessible concurrently whilst flushing happens - can end up reindexing deleted data (i.e. I remove an item from the cache, but by the time I commit the batch, someone read it and pulled it back in).

In this case, should we perhaps leave the cache invalidation to *after* the batch flush? It should be quick since at that point it's just in-memory ops, and if anyone attempt to re-read a deleted item, it will access the correct thing from disk (or nothing in this casE)